bionic/core/task_execution.py

```diff
@@ -701,25 +701,65 @@ def _check_accessor_for_version_problems(self):
+                assert old_prov.nominally_matches(new_prov)
+                # If the bytecode doesn't match, we found the problematic pair.
+                if old_prov.bytecode_hash != new_prov.bytecode_hash:
+                    message = f"""
+                        Found a cached artifact with the same descriptor
+                        ({self._cache_accessor.query.dnode.to_descriptor()!r})
+                        and version (major={old_prov.code_version_major!r},
+                        minor={old_prov.code_version_minor!r}),
+                        but created by different code.
+                        It appears that the code function that outputs
+                        {new_prov.descriptor!r}
+                        was changed (old bytecode hash {old_prov.bytecode_hash!r};
+                        new bytecode hash {new_prov.bytecode_hash!r})
+                        but the function's version number was not.
+                        Change @version(major=) to indicate that your
+                        function's behavior has changed, or @version(minor=)
+                        to indicate that it has *not* changed.
+                        """
+                    raise CodeVersioningError(oneline(message))
+                # If the provenances nominally match, they must have essentially the
+                # same structure.
+                assert len(old_prov.dep_digests) == len(new_prov.dep_digests)
+                # Since these provenances match nominally and have matching bytcode,
+                # the mismatch must be in one of their dependencies. We'll iterate
+                # through them to figure out which one.
+                for old_dep_digest, new_dep_digest in zip(
+                    old_prov.dep_digests, new_prov.dep_digests
+                ):
+                    # If this digest pair matches, it must not be where the problem is.
+                    if old_dep_digest.exact_hash == new_dep_digest.exact_hash:
+                        continue
+
+                    # Not all digests have provenances, but these should. Digests of
+                    # non-persisted values have provenances, and if these were persisted
+                    # then their exact hashes would be the same as their nominal hashes,
+                    # so they would have matched above.
+                    old_dep_prov = old_dep_digest.provenance
+                    new_dep_prov = new_dep_digest.provenance
+                    locate_mismatched_provenances_and_raise(old_dep_prov, new_dep_prov)
+                assert False
+
+            try:
+                locate_mismatched_provenances_and_raise(old_prov, new_prov)
+            except AssertionError as e:
+                message = """
+                Unable to complete assisted versioning check due to a bug in Bionic;
+                please report this to the developers
+                """
+                logger.warn(oneline(message), exc_info=e)
```

Comment:
Should this be an error instead? If there is a bug, the users cannot rely on assisted versioning anyway. Plus this warning is going to spam the users since we refresh the cache accessors a few times too (and all the descendents would log this error too).

Reply:
Good question. I see your point, but here was my thinking:

1. Assisted versioning is already best-effort; users can't rely on it to be 100% accurate. Also, it's just a backstop to catch you if you forget to manually update a version.
2. If this does happen, there's probably still a versioning error; we just can't figure out exactly where it is. (Maybe we should make this more clear in the message.)
3. With the warning, the user has the option of filtering the warning and continuing; with an exception they will have to just give up.

Maybe I could expand the message to explain the user's options here (try to fix the error, clear the cache, or just stop using assisted versioning)?

Reply:
The updated message looks good!